### PR TITLE
HEP: support RWX volume to downstream cluster

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
Add HEP for RWX volume support

**Solution:**
Add HEP for RWX volume support

**Related Issue:**
https://github.com/harvester/harvester/issues/1992

### CODE REVIEW

**Comment by Vicente-Cheng** [comment]
cc @c3y1huang @derekbit, I could not add you to the reviewer list, but please help to check this. Thanks!

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+A quick overview of the design could be checked below:
+
+![Downstream RWX Support](20240523-support-rwx-volume-to-downstream-cluster/general-rwx.png)
+
+### The best solution
+
+Like the RWO solution, we would like to leverage the libvirt capability to hotplug the filesystem and then use the workload pod as a host mountpoint, but currently, it is not supported. The hotplug filesystem on the kubevirt side is referred to as virtiofs. However, virtiofs does not support live migration, so we could not use this solution.
+
+Related feature request to Kubevirt: https://github.com/kubevirt/kubevirt/issues/11699
```
Comment (WebberHuang1118) [inline enhancements/20240523-support-rwx-volume-to-downstream-cluster.md:59]: Since Kubevirt doesn't support filesystem mode for hotplug volume https://github.com/kubevirt/kubevirt/issues/11699, so we plan to use the filesystem created by Longhorn and export it to the guest cluster?

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+#### 2. The Backup Store of MongoDB
+MongoDB could configure the backup store with the Kubernetes Operator. The prerequisite is the RWX volume, which is similar to the NueVector. The user needs this feature once they want to use K8s Operator to backup MongoDB.
+
+Reference: https://www.mongodb.com/docs/kubernetes-operator/v1.25/tutorial/configure-file-store/
+
+### User Experience In Detail
+
+With this feature, the user could create volumes with two options in the downstream cluster. The first is the RWO volume. The second is the RWX volume. Then user could attach use the RWX volume as previous RWO volume. The Harvester CSI driver would handle the internal work.
+The UX did not change too much. The only difference is the user needs to select the RWX volume.
```
Comment (WebberHuang1118) [inline enhancements/20240523-support-rwx-volume-to-downstream-cluster.md:42]: IIUC, the current RWO volume is the filesystem formatted on the downstream cluster VM, and the LH RWX volumes leverage the filesystem created by LH share manager. How will we transfer the previous RWO filesystem to the RWX filesystem? Can you explain more?

**Review by WebberHuang1118 — CHANGES_REQUESTED** [review]
LGTM, just a few questions to confirm my understanding is correct, thanks.

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md (hunk 1 above)
Comment (Vicente-Cheng) [reply enhancements/20240523-support-rwx-volume-to-downstream-cluster.md:59]: This is a different issue. We would like to provide the RWX volume on v1.4. The longhorn RWX volume is our built-in solution. It provides the RWX volume with the NFS service. From the user view, this is the filesystem we can use directly. So that why we check the filesystem hotplug/Live migrate capability on the kubevirt.

TL;DR, Kubevirt can hotplug/live migrate for the block device but can't for the filesystem. So, we need a new mechanism to provide RWX volume to the downstream cluster.

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md (hunk 2 above)
Comment (Vicente-Cheng) [reply enhancements/20240523-support-rwx-volume-to-downstream-cluster.md:42]: Yeah, as you know, we will format the RWO (block) volume when we provide the filesystem view on the k8s workload.

For the RWX volume, we will use another code path to avoid the format and directly mount the NFS endpoint.
Like below:
RWO volume: format -> mount
RWX volume: mount

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+The whole flow might be like that when get the request to create RWX volume, Harvester CSI Driver need to create corresponding RWX volume on Longhorn. Then, the Harvester CSI driver will create a global path on NodeStage. Last, with NodePublish, the Harvester CSI driver will bind mount the global path for the workload.
+
+#### 3. NetworkFS controller on Harvester (New Added)
+We need to create a workload pod to bring up the Longhorn-share manager so that the Longhorn-share-manager can create the NFS endpoint.
+We would like to introduce the NetworkFS controller to handle(create) the pod. The advantage with this component is that we could control the whole network filesystem on Harvester side. The network filesystem could be NFS. Later, we could add more network filesystem like CIFS(SMB)/CephFS. Also, we can add some network strategies like network policy or service endpoint to limit the access to the network filesystem.
```
Comment (WebberHuang1118) [inline enhancements/20240523-support-rwx-volume-to-downstream-cluster.md]: Can you explain more about how to expose the LH NFS endpoint to the downstream cluster during CSI `NodeStage`? Does it require the LH feature [feat(LEP): storage network for RWX volumes](https://github.com/longhorn/longhorn/pull/8616)?

Thanks

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+Create an endpoint for Harvester VM to use to have the shared filesystem.
+
+## Proposal
+
+The basic idea is to use the Longhorn RWX volume. It provides the NFS endpoint through the Longhorn share manager, so we need to focus on how the downstream cluster interacts with NFS. Once the downstream cluster interacts with this NFS endpoint, we can attach the RWX volume to any workload through the Harvester CSI driver.
+
+### User Stories
+
+#### 1. The Configuration Backup of NueVector
```
Comment (asettle) [inline enhancements/20240523-support-rwx-volume-to-downstream-cluster.md]: NeuVector (just fyi)

Comment (Vicente-Cheng) [reply enhancements/20240523-support-rwx-volume-to-downstream-cluster.md]: updated!

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md (hunk 3 above)
Comment (Vicente-Cheng) [reply enhancements/20240523-support-rwx-volume-to-downstream-cluster.md]: Update the networkfs-manager and the responsibility of its CRD for the endpoint information.

**Comment by Vicente-Cheng** [comment]
Hi guys, please help to review it again.

I have some updates for the above discussion.

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+### Implementation Overview
+
+A quick overview of the design could be checked below:
+
+![Downstream RWX Support](20240523-support-rwx-volume-to-downstream-cluster/general-rwx.png)
+
+### Solution 1.
+
+Like the RWO solution, we would like to leverage the libvirt capability to hotplug the filesystem and then use the workload pod as a host mountpoint, but currently, it is not supported. The hotplug filesystem on the kubevirt side is referred to as virtiofs. However, virtiofs does not support live migration, so we could not use this solution.
```
Comment (innobead) [inline enhancements/20240523-support-rwx-volume-to-downstream-cluster.md:57]: This should be the general approach if we want to support this for 3rd storage eventually.

File: enhancements/20240523-support-rwx-volume-to-downstream-cluster.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+#### 1. Longhorn
+Longhorn share-manager can provide the NFS endpoint through the storage network.
+
+Related issue: https://github.com/longhorn/longhorn/issues/8184
+
+#### 2. Harvester CSI Driver
+Harvester CSI Driver need to add capability to create the RWX volume through Longhorn client. Also, we should add the corresponding handles on NodeStage/NodePublish stage. The NodeStage will mount the NFS endpoint to the host as global path. Then, the NodePublish will bind mount the global path for the workload.
+
+The whole flow might be like that when get the request to create RWX volume, Harvester CSI Driver need to create corresponding RWX volume on Longhorn. Then, the Harvester CSI driver will create a global path on NodeStage. Last, with NodePublish, the Harvester CSI driver will bind mount the global path for the workload.
```
Comment (ihcsim) [inline enhancements/20240523-support-rwx-volume-to-downstream-cluster.md:87]: > Harvester CSI Driver need to create corresponding RWX volume on Longhorn

Is this limited to LH only? What will it take to make this more generic? Theoretically, the Harvester CSI driver can create RWX volumes of any storage class, and it's up to the Harvester-side CSI drivers to do the right thing, right?
